### PR TITLE
Use ipfsd-ctl to create test IPFS instances

### PR DESCRIPTION
**Pull request by haadcode**
This PR brings back [ipfsd-ctl](https://github.com/ipfs/js-ipfsd-ctl) to create test IPFS instances.

In addition, this PR will move a lot of the boilerplate test setup code to it's on test utils as well as fix some basic tests for the stores and remove couple of obsolete/unused public methods from OrbitDB class.

### CODE REVIEW

File: test/replicate.test.js [hunk 1]
```diff
@@ -4,293 +4,197 @@ const assert = require('assert')
...
-      if (db1)
-        await db1.drop()
-      
-      if (db2) 
-        await db2.drop()
-
-      if(orbitdb1) 
-        await orbitdb1.stop()
+Object.keys(testAPIs).forEach(API => {
+  describe.only(`orbit-db - Replication (${API})`, function() {
```
Comment (thiagodelgado111) [inline test/replicate.test.js]: Not sure if this is still WIP but there's a `.only` here that will avoid other tests from running on CI

**Comment by thiagodelgado111** [comment]
I'm getting an error log while running tests (this time with automatic replication) locally. Maybe it's just a negative test failing as it should and I missed it?

```
Error: underlying socket has been closed
    at /orbit-db/node_modules/multiplex/src/index.js:477:37
    at Array.forEach (<anonymous>)
    at Multiplex.destroy (/orbit-db/node_modules/multiplex/src/index.js:475:10)
    at MultiplexMuxer.end (/orbit-db/node_modules/libp2p-multiplex/src/muxer.js:49:20)
    at each (/orbit-db/node_modules/libp2p-swarm/src/index.js:127:20)
    at /orbit-db/node_modules/async/internal/withoutIndex.js:9:16
    at replenish (/orbit-db/node_modules/async/internal/eachOfLimit.js:64:17)
    at /orbit-db/node_modules/async/internal/eachOfLimit.js:68:9
    at eachOfLimit (/orbit-db/node_modules/async/eachOfLimit.js:39:36)
    at /orbit-db/node_modules/async/internal/doLimit.js:9:16
    at exports.default (/orbit-db/node_modules/async/eachOf.js:9:5)
    at eachLimit (/orbit-db/node_modules/async/each.js:80:24)
    at series (/orbit-db/node_modules/libp2p-swarm/src/index.js:126:15)
    at /orbit-db/node_modules/async/internal/parallel.js:31:39
    at replenish (/orbit-db/node_modules/async/internal/eachOfLimit.js:64:17)
    at /orbit-db/node_modules/async/internal/eachOfLimit.js:68:9
    at eachOfLimit (/orbit-db/node_modules/async/eachOfLimit.js:39:36)
    at /orbit-db/node_modules/async/internal/doLimit.js:9:16
    at _parallel (/orbit-db/node_modules/async/internal/parallel.js:30:5)
    at series (/orbit-db/node_modules/async/series.js:83:26)
    at Swarm.close (/orbit-db/node_modules/libp2p-swarm/src/index.js:125:5)
    at series (/orbit-db/node_modules/libp2p/src/index.js:240:26)
    at /orbit-db/node_modules/async/internal/parallel.js:31:39
    at replenish (/orbit-db/node_modules/async/internal/eachOfLimit.js:64:17)
    at iterateeCallback (/orbit-db/node_modules/async/internal/eachOfLimit.js:49:17)
    at /orbit-db/node_modules/async/internal/onlyOnce.js:12:16
    at /orbit-db/node_modules/async/internal/parallel.js:36:13
    at series (/orbit-db/node_modules/libp2p/src/index.js:238:9)
    at /orbit-db/node_modules/async/internal/parallel.js:31:39
    at replenish (/orbit-db/node_modules/async/internal/eachOfLimit.js:64:17)
    at /orbit-db/node_modules/async/internal/eachOfLimit.js:68:9
    at eachOfLimit (/orbit-db/node_modules/async/eachOfLimit.js:39:36)
    at /orbit-db/node_modules/async/internal/doLimit.js:9:16
    at _parallel (/orbit-db/node_modules/async/internal/parallel.js:30:5)
    at series (/orbit-db/node_modules/async/series.js:83:26)
    at Node.stop (/orbit-db/node_modules/libp2p/src/index.js:233:5)
    at Function.stop.promisify (/orbit-db/node_modules/ipfs/src/core/components/libp2p.js:66:24)
    at Object.stop (/orbit-db/node_modules/promisify-es6/index.js:32:27)
    at series (/orbit-db/node_modules/ipfs/src/core/components/stop.js:35:27)
    at /orbit-db/node_modules/async/internal/parallel.js:31:39
    at replenish (/orbit-db/node_modules/async/internal/eachOfLimit.js:64:17)
    at iterateeCallback (/orbit-db/node_modules/async/internal/eachOfLimit.js:49:17)
    at /orbit-db/node_modules/async/internal/onlyOnce.js:12:16
    at /orbit-db/node_modules/async/internal/parallel.js:36:13
    at asyncEach (/orbit-db/node_modules/libp2p-floodsub/src/index.js:284:7)
    at /orbit-db/node_modules/async/internal/once.js:12:16
    at iterateeCallback (/orbit-db/node_modules/async/internal/eachOfLimit.js:47:24)
    at /orbit-db/node_modules/async/internal/onlyOnce.js:12:16
    at setImmediate (/orbit-db/node_modules/libp2p-floodsub/src/peer.js:182:7)
    at Immediate._onImmediate (/orbit-db/node_modules/async/internal/setImmediate.js:27:16)
    at runCallback (timers.js:756:18)
    at tryOnImmediate (timers.js:717:5)
    at processImmediate [as _immediateCallback] (timers.js:697:5)
```

**Comment by haadcode** [comment]
The tests are finally passing again. 

I added a workaround for the "Error: underlying socket has been closed" error, as it seems really weird. See here https://github.com/orbitdb/orbit-db/pull/335/files#diff-11b1d30de414c2bebf3a03b5f2bd1b73R48 and here https://github.com/orbitdb/orbit-db/pull/335/files#diff-b1d58ced77879cec2a109423fb3755a8R87. Not sure what's happening but it may not be related to libp2p because if we comment out the leveldb's method to destroy the files from the local filesystem (https://github.com/orbitdb/orbit-db-cache/blob/master/Cache.js#L59 which gets called from https://github.com/orbitdb/orbit-db-store/blob/master/src/Store.js#L171), the error is gone 😕❓❓❓😕

Otherwise, this is ready for merge but would love another round of review to catch any issues!

**Comment by thiagodelgado111** [comment]
I have no idea :/ Pulled your last changes, reinstalled everything and now tests are passing. Previously it was failing intermittently, so it does look like a race condition situation

**Comment by haadcode** [comment]
Found the bug and it was indeed a race condition. We were not waiting for unsubscribing from pubsub internally and that caused the race (receiving messages after closing the database). Fixed it and updated the PR and removed the timeout workarounds from the test.

Ready to merge.

**Comment by haadcode** [comment]
Thank @thiagodelgado111 and everyone else who tested and reviewed this PR! ❤️ It's now in master and would love some more testing. 

Will be merging https://github.com/orbitdb/orbit-db/pull/328 and then make a new release (0.20.0). Hopefully this week.

**Pull request merged**